config-cmd/signin.go

```diff
@@ -0,0 +1,127 @@
+will redirect you to localhost where this command will be listening
+and save the temporary credentials to local configuration file.`,
+		RunE: cmdSignin,
+	}
+	cmd.Flags().IntP("port", "p", 0, "Port to use; defaults to random ephemeral port.")
+
+	root.Command.AddCommand(cmd)
+
+	config.RegisterOptions("signin", map[string]config.OptionDefinition{
+		"loginUrl": config.OptionDefinition{
+			Description: "URL for the login service.",
+			Default:     "https://login.taskcluster.net",
+			Validate:    isString,
+		},
+	})
+}
+
+func cmdSignin(cmd *cobra.Command, _ []string) error {
+	// Load configuration
+	fmt.Fprintln(cmd.OutOrStdout(), "Starting")
+
+	// Find port, choose 0 meaning random port, if none
+	port, _ := cmd.Flags().GetInt("port")
+
+	// Setup server that we can shutdown gracefully
+	s := graceful.Server{
+		Timeout: 5 * time.Second,
+		Server:  &http.Server{},
+	}
+
+	// Handle callback
+	var serr error
+	s.Server.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		qs := r.URL.Query()
+		config.Configuration["config"]["clientId"] = qs.Get("clientId")
+		config.Configuration["config"]["accessToken"] = qs.Get("accessToken")
+		config.Configuration["config"]["certificate"] = qs.Get("certificate")
+
+		serr = config.Save(config.Configuration)
+		if serr == nil {
+			fmt.Fprintln(cmd.OutOrStdout(), "Credentials saved in configuration file.")
+		}
+
+		title := "Successful"
+		result := "You have successfully signed in."
+		if serr != nil {
+			title = "Failed"
+			result = "Failed to save credentials!"
+		}
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(`
+			<!doctype html>
+			<html>
+				<head>
+					<title>Sign-In ` + title + `</title>
+					<script>
+						// Close the window after 500ms
+						setTimeout(function() {
+							window.close();
+						}, 500);
+					</script>
+				</head>
+				<body>
+					<h1>` + result + `.</h1>
+				</body>
+			</html>
+		`))
+		s.Stop(50 * time.Millisecond)
+	})
+
+	// Start listening on localhost
+	listener, err := net.ListenTCP("tcp", &net.TCPAddr{
+		IP:   []byte{127, 0, 0, 1},
+		Port: port,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to listen on localhost, error: %s\n", err)
+	}
+
+	// Construct URL for login service and open it
+	target := "http://" + strings.Replace(listener.Addr().String(), "127.0.0.1", "localhost", 1)
+	description := url.QueryEscape(
+		`Login and click the "Grant Access" button to transfer
+		temporary credentials to the taskcluster CLI client`,
+	)
+	loginURL := config.Configuration["signin"]["loginUrl"].(string)
+	loginURL += "/?target=" + url.QueryEscape(target) + "&description=" + description
+
+	// Open browser
+	fmt.Fprintln(cmd.OutOrStdout(), "Listening for a callback on: "+target)
+	webbrowser.Open(loginURL, webbrowser.NewWindow, true)
+
+	// Start serving
+	s.Serve(listener)
+
+	if serr != nil {
+		return fmt.Errorf("failed to save configuration, error: %s\n", serr)
+	} else {
+		return nil
```

Comment:
nit: if block ends with a return statement, so drop this else and outdent its block (golint)